xarray/tests/test_backends.py

```diff
@@ -7260,6 +7267,77 @@ def test_zarr_entrypoint(tmp_path: Path) -> None:
     assert not entrypoint.guess_can_open("something.zarr.txt")
 
 
+@requires_h5netcdf
+@requires_netCDF4
+@requires_pydap
+@requires_zarr
+def test_remote_url_backend_auto_detection() -> None:
+    """
+    Test that remote URLs are correctly selected by the backend resolution system.
+
+    This tests the fix for issue where netCDF4, h5netcdf, and pydap backends were
+    claiming ALL remote URLs, preventing remote Zarr stores from being
+    auto-detected.
+
+    See: https://github.com/pydata/xarray/issues/10801
+    """
+    from xarray.backends.plugins import guess_engine
+
+    # Test cases: (url, expected_backend)
+    test_cases = [
+        # Remote Zarr URLs
+        ("https://example.com/store.zarr", "zarr"),
+        ("http://example.com/data.zarr/", "zarr"),
+        ("s3://bucket/path/to/data.zarr", "zarr"),
+        # Remote netCDF URLs (non-DAP) - h5netcdf wins (first in order, no query params)
+        ("https://example.com/file.nc", "h5netcdf"),
+        ("http://example.com/data.nc4", "h5netcdf"),
+        ("https://example.com/test.cdf", "h5netcdf"),
+        ("s3://bucket/path/to/data.nc", "h5netcdf"),
```

Comment:
this will need updating after https://github.com/pydata/xarray/pull/10817, sorry!